src/core/processing/qgsprocessingutils.cpp

```diff
@@ -721,7 +721,19 @@ QVariant QgsProcessingFeatureSource::maximumValue( int fieldIndex ) const
   return mSource->maximumValue( fieldIndex );
 }
 
-QgsFeatureSource *QgsProcessingFeatureSource::source() const
+QgsExpressionContext QgsProcessingFeatureSource::createExpressionContext( const QgsProcessingContext &context ) const
 {
-  return mSource;
+  QgsExpressionContext expressionContext;
+  QgsExpressionContextGenerator *generator = dynamic_cast<QgsExpressionContextGenerator *>( mSource );
+  if ( generator )
+  {
+    expressionContext = generator->createExpressionContext();
+  }
+  else
+  {
+    expressionContext
```

Comment:
I don't think this is needed - if the source isn't a generator, then adding these here will result in duplicate global/project scopes in the context (since they'll already be present in the context's original expression context)

Reply:
I was more thinking about this being an independent piece from which a consumer expects some context.
I'd prefer that this delivers a context and then dismiss anything that some python script might have pre-populated. If python needs to add something it should do it after.

Reply:
Ok, that's a noble goal. But in this case the context's expression context may have other scopes also - such as a model scope. Adding the global/project scope after these breaks the cascading design of scopes.

Reply:
Isn't that added after already?

```
QgsExpressionContext QgsProcessingAlgorithm::createExpressionContext( const QVariantMap &parameters,
    QgsProcessingContext &context, QgsProcessingFeatureSource *source ) const
{
  // start with context's expression context
  QgsExpressionContext c = context.expressionContext();

  // If there's a source capable of generating a context scope, use it
  if ( source )
  {
    const auto &scopes = source->createExpressionContext( context ).takeScopes();
    for ( QgsExpressionContextScope *scope : scopes )
      c << scope;
  }
  else if ( c.scopeCount() == 0 )
  {
    //empty scope, populate with initial scopes
    c << QgsExpressionContextUtils::globalScope()
      << QgsExpressionContextUtils::projectScope( context.project() );
  }

  // !! HERE !!
  c << QgsExpressionContextUtils::processingAlgorithmScope( this, parameters, context );
  return c;
}
```

Reply:
I'm referring to a parent model scope (which incidentally i could have sworn i'd already added, but turns out i haven't yet!) model scope should be used to expose results from previous child algorithms to expressions in the child algorithm, and raw input parameter values too). It should then ideally be:

- global
- project
- model
- algorithm
- source

(or source/algorithm. not sure on those last two - it probably doesn't matter). But the model scope would sit above the alg/source scope.

If you're concerned about relying on the python generation of the processing context's initial expression context, we could easily move that to c++ and have it so:
- creating a new QgsProcessingContext = initially populate the context's expression context with a global scope
- setting the project for the context = adding the project scope

Reply:
What I would either propose is

```
/**
 * Returns a fully populated expression context with global/project/model/alg/source scopes
 * Based on the processing context in the parameter.
 */
QgsExpressionContext QgsProcessingFeatureSource( QgsProcessingContext )
```

or

```
/**
 * Returns a expression context scope for this source (might be a layer scope)
 */
QgsExpressionContextScope *QgsProcessingFeatureSource()
```

Reply:
My vote goes to the second - seems a bit more flexible to me.

Reply:
Mine too... Just, what we have so far is only a `QgsExpressionContextGenerator` interface and no `QgsExpressionContextScopeGenerator` interface. Taking the last scope of the generated context might work but not really what I'd like to do... So, this might require to add the second interface to QgsVectorLayer as well then...